service/src/main/java/com/commercetools/pspadapter/payone/PaymentHandler.java

```diff
@@ -55,68 +52,90 @@ public PaymentHandler(String payoneInterfaceName, String tenantName,
-                    final PaymentWithCartLike paymentWithCartLike = commercetoolsQueryExecutor.getPaymentWithCartLike(paymentId);
-                    String paymentInterface = paymentWithCartLike.getPayment().getPaymentMethodInfo().getPaymentInterface();
-                    if (!payoneInterfaceName.equals(paymentInterface)) {
-                        logger.warn(tenantNameKeyValue, "Wrong payment interface name: expected [{}], found [{}]", payoneInterfaceName, paymentInterface);
-                        return new PaymentHandleResult(HttpStatusCode.BAD_REQUEST_400);
-                    }
-
-                    paymentDispatcher.dispatchPayment(paymentWithCartLike);
-                    return new PaymentHandleResult(HttpStatusCode.OK_200);
-                } catch (final ConcurrentModificationException cme) {
-                    logger.warn(tenantNameKeyValue, "Exception on dispatchPayment with id [{}]. Retry {} of {}", paymentId, i + 1, RETRIES_LIMIT);
-                    Thread.sleep(RETRY_DELAY);
-                }
-            }
-
-            logger.error(
-                tenantNameKeyValue,
-                "The payment [{}] couldn't be processed after {} retries", paymentId, RETRIES_LIMIT);
-            return new PaymentHandleResult(HttpStatusCode.ACCEPTED_202,
-                    format("The payment couldn't be processed after %s retries", RETRIES_LIMIT));
-
+            return processPaymentOrRetry(paymentId, 0);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            return handleConcurrentModificationException(paymentId, concurrentModificationException);
         } catch (final NotFoundException | NoCartLikeFoundException e) {
             return handleNotFoundException(paymentId, e);
         } catch (final ErrorResponseException e) {
             return errorResponseHandler(e, paymentId);
-        } catch (final CompletionException e) {
-            return completionExceptionHandler(e, paymentId);
         } catch (final Exception e) {
-            return handleThrowableInResponse(e, paymentId);
+            return handleException(e, paymentId);
         }
     }
 
-     private PaymentHandleResult handleNotFoundException(@Nonnull String paymentId, @Nonnull Exception exception ) {
+    private PaymentHandleResult processPaymentOrRetry(
+        @Nonnull final String paymentId,
+        final int currentAttemptCount) throws InterruptedException {
+
+        try {
+            return processPayment(paymentId);
+        } catch (final ConcurrentModificationException concurrentModificationException) {
+            if (currentAttemptCount == RETRIES_LIMIT - 1) {
+                throw concurrentModificationException;
+            } else {
+                Thread.sleep(RETRY_DELAY);
+                return processPaymentOrRetry(paymentId, currentAttemptCount + 1);
+            }
+        }
+    }
+
+    private PaymentHandleResult processPayment(@Nonnull final String paymentId) {
```

Comment:
Maybe add a throw declaration here.

Reply:
yep, I agree. Even though we should not declare a throwing an unchecked exception. But in this case, it's a good hint for the called of the method that they should catch it.

Reply:
https://github.com/commercetools/commercetools-payone-integration/pull/257/commits/1c02b6420fc08be0586c362b1e48ba950f2d338a